internal/rsl/rsl.go

```diff
@@ -0,0 +1,296 @@
+}
+
+// Commit creates a commit object in the RSL for the Annotation.
+func (a Annotation) Commit(sign bool) error {
+	repo, err := common.GetRepositoryHandler()
+	if err != nil {
+		return err
+	}
+
+	// Check if referred entries exist in the RSL namespace.
+	for _, id := range a.RSLEntryIDs {
+		if _, err := GetEntry(id); err != nil {
+			return err
+		}
+	}
+
+	message, err := a.createCommitMessage()
+	if err != nil {
+		return err
+	}
+
+	return gitinterface.Commit(repo, plumbing.ZeroHash, RSLRef, message, sign)
+}
+
+func (a Annotation) createCommitMessage() (string, error) {
+	lines := []string{
+		AnnotationHeader,
+		"",
+	}
+
+	for _, entry := range a.RSLEntryIDs {
+		lines = append(lines, fmt.Sprintf("%s: %s", EntryIDKey, entry.String()))
+	}
+
+	if a.Skip {
+		lines = append(lines, fmt.Sprintf("%s: true", SkipKey))
+	} else {
+		lines = append(lines, fmt.Sprintf("%s: false", SkipKey))
+	}
+
+	if len(a.Message) != 0 {
+		var message strings.Builder
+		messageBlock := pem.Block{
+			Type:  AnnotationMessageBlockType,
+			Bytes: []byte(a.Message),
+		}
+		if err := pem.Encode(&message, &messageBlock); err != nil {
+			return "", err
+		}
+		lines = append(lines, strings.TrimSpace(message.String()))
+	}
+
+	return strings.Join(lines, "\n"), nil
+}
+
+// GetLatestEntry returns the latest entry available locally in the RSL.
+// TODO: There is no information yet about the signature for the entry.
+func GetLatestEntry() (EntryType, error) {
+	repo, err := common.GetRepositoryHandler()
+	if err != nil {
+		return nil, err
+	}
+
+	ref, err := repo.Reference(plumbing.ReferenceName(RSLRef), true)
+	if err != nil {
+		return nil, err
+	}
+
+	commitObj, err := repo.CommitObject(ref.Hash())
+	if err != nil {
+		return nil, err
+	}
+
+	return parseRSLEntryText(commitObj.Message)
+}
+
+// GetEntry returns the entry corresponding to entryID.
+// TODO: There is no information yet about the signature for the entry.
+func GetEntry(entryID plumbing.Hash) (EntryType, error) {
+	repo, err := common.GetRepositoryHandler()
+	if err != nil {
+		return nil, err
+	}
+
+	ref, err := repo.Reference(plumbing.ReferenceName(RSLRef), true)
+	if err != nil {
+		return nil, err
+	}
+
+	iteratorHash := ref.Hash()
+	for {
+		commitObj, err := repo.CommitObject(iteratorHash)
+		if err != nil {
+			return nil, err
+		}
+
+		if iteratorHash == entryID {
+			return parseRSLEntryText(commitObj.Message)
+		}
+
+		if len(commitObj.ParentHashes) == 0 {
+			return nil, ErrRSLEntryNotFound
+		}
+
+		if len(commitObj.ParentHashes) > 1 {
+			return nil, ErrRSLBranchDetected
+		}
+
+		iteratorHash = commitObj.ParentHashes[0]
+	}
+}
+
+func parseRSLEntryText(text string) (EntryType, error) {
+	text = strings.TrimSpace(text)
+	if strings.HasPrefix(text, AnnotationHeader) {
+		return parseAnnotationText(text)
+	}
+	return parseEntryText(text)
+}
+
+func parseEntryText(text string) (*Entry, error) {
+	lines := strings.Split(text, "\n")
+	lines = lines[2:]
+	if len(lines) < 2 {
```

Comment:
This needs to happen before the access above, otherwise the lines[2] might be OOB

Reply:
Good catch! I moved it one line up and changed the comparison to < 4 to account for the header. I also added two tests apiece for entries and annotations, one with missing header, the other with missing information but including the header to assert `ErrInvalidRSLEntry`